File: avl_tree/AVLTree.java
import java.util.*;


class AVLNode {
    public AVLNode right;
    public AVLNode left;
    public int data;
    public int height;
    AVLNode(int num) {
        data = num;
        left = null;
        right = null;
        height = 1; 
    }
    
    public void setLeft(AVLNode temp) {
        left = temp;
    }
    
    public void setRight(AVLNode temp) {
        right = temp;
    }
}

class AVL{
    private AVLNode root;
    public AVL() {
        root = null;
    }
    
    public int height(AVLNode root) {
        if(root == null)
            return 0; 
        return root.height; 
    }

    public boolean isEmpty() {
        return root == null;
    }
    
    public int getBalance(AVLNode root) {
        if(root == null)
            return 0;
        return height(root.left) - height(root.right); 
    }

    public void insert(int data) {
        root = insert(root, data);
    }
    
    private AVLNode rightRotate(AVLNode root) {
        AVLNode y = root.left;  
        AVLNode temp = y.right;
        y.right = root;
        root.left = temp;
        root.height= 1 + Math.max(height(root.left), height(root.right));    
        y.height = 1 + Math.max(height(y.left), height(y.right));   
        return y;
    } 
      
    private AVLNode leftRotate(AVLNode root) {
        AVLNode y = root.right;  
        AVLNode temp = y.left;
        y.left = root;
        root.right = temp;
        root.height= 1 + Math.max(height(root.left), height(root.right));    
        y.height = 1 + Math.max(height(y.left), height(y.right));   
        return y;
    } 
    
    private AVLNode insert(AVLNode root, int data) {
        if (root == null) {
            root = new AVLNode(data);
        }
        else{   
              if (data < root.data) { 
              root.left = insert(root.left, data);
            }
              else if(data > root.data) {
                root.right = insert(root.right, data);
            } 
              else
                return root;
        }
        root.height = 1 + Math.max(height(root.right), height(root.left));  
        int balance = height(root.left) - height(root.right); 
            
        if(balance > 1 && getBalance(root.left) >= 0)  //left-left case
            return rightRotate(root);
        
        else if(balance < -1 && getBalance(root.right) < 0)  //right-right case 
            return leftRotate(root);
        
        else if(balance > 1 && getBalance(root.right) < 0) {   //left-right case
            root.left = leftRotate(root.left);   
            return rightRotate(root);
        }
        
        else if(balance < -1 && getBalance(root.left) >= 0) {  //right-left case 
            root.right = rightRotate(root.right);
            return leftRotate(root);   
        } 
        return root;
    }
    
    public boolean search(int val) { 
        return search(root, val);
    }
    
    private boolean search(AVLNode root, int val) {
        boolean found = false;
        while ((root != null) && !found) {
            int rval = root.data;
            if (val < rval)
                root = root.left;
            else if (val > rval)
                root = root.right;
            else{
                found = true;
                break;
            }
            found = search(root, val);
        }
        return found;
    }
     
    public void inorder() {
        inorder(root);
    }
     
    private void inorder(AVLNode root) {
        if (root != null) {
            inorder(root.left);
            System.out.print(root.data + " ");
            inorder(root.right);
        }
    }
    public void delete(int k) {
        if (isEmpty())
            System.out.println("Tree Empty");
        else if (!search(k))
            System.out.println("Sorry " + k + " is not present");
        else{
            root = delete(root, k);
            System.out.println(k + " deleted from the tree");
        }
    }
    private AVLNode delete(AVLNode root, int k) {
        AVLNode p;
        AVLNode n;
        if (root.data == k) {
            AVLNode left; 
            AVLNode right;
            left = root.left;
            right = root.right;
            if (left == null && right == null)
                return null;
            else if (left == null)
                return right;
            else if (right == null)
                return left;
            else{
                p = right;
                while (p.left != null) {
                    p = p.left;
                }
                root.data = p.data; 
                root.right= delete(right, p.data);
            }
        }
        else if (k < root.data) {
            n = delete(root.left, k);
            root.setLeft(n);
        }
        else{
            n = delete(root.right, k);
            root.setRight(n);             
        }
            
        if(root == null)
            return root;
            root.height = 1 + Math.max(height(root.right), height(root.left));   
            int balance = height(root.left) - height(root.right); 
            
        if(balance > 1 && getBalance(root.left) >= 0)  //left-left case
            return rightRotate(root);
        
        else if(balance < -1 && getBalance(root.right) < 0)  //right-right case 
            return leftRotate(root);
        
        else if(balance > 1 && getBalance(root.right) < 0) {   //left-right case
            root.left = leftRotate(root.left);   
            return rightRotate(root);
        }
        else if(balance < -1 && getBalance(root.left) >= 0) {  //right-left case 
            root.right = rightRotate(root.right);
            return leftRotate(root);   
        } 
        return root;
    }
}

public class AvlTree{
    public static void main(String[] args) {
        AVL s = new AVL();        
        s.insert(10);
        s.insert(14);
        s.insert(5);
        s.insert(15);
        s.insert(16);
        s.insert(2);
        s.insert(1);
        s.insert(7);
        s.inorder();    // Printing in inorder
        System.out.println(s.search(7));    // Searching 7 
        s.delete(1);    // Deleting 1
        System.out.println(s.search(1));    // Searching 1
        s.inorder();    // Printing in inorder, again 
    }
}

File: avl_tree/avl_tree.c
// Implements an AVL tree in C

#include <stdlib.h>
#include <stdio.h>

#define max(a,b) (((a)>(b))?(a):(b))

typedef struct AVLNode {
	struct AVLNode* left;		// left child node
	struct AVLNode* right;		// right child node
	struct AVLNode* parent;		// parent node
	int value; 					// integer type data
	int height;					// current height in the tree
} AVLNode;

typedef struct AVLTree {
	struct AVLNode* root;		// the tree's root node
	int size;					// the tree's size
} AVLTree;


// Adjusts the height of the given node respective to its current position
void adjust_height(AVLNode* avln) {
	if (avln->left == NULL && avln->right == NULL) {
		avln->height = 1;
	} else if (avln->left == NULL) {
		avln->height = 1 + avln->right->height;
	} else if (avln->right == NULL) {
		avln->height = 1 + avln->left->height;
	} else {
		avln->height = 1 + max(avln->left->height, avln->right->height);
	}
}

// Performs a left rotation on the given node in the given tree
void rotate_left(AVLTree* avlt, AVLNode* avln) {
	AVLNode* y = avln->right;
	avln->right = y->left;

	if (y->left != NULL) {
		y->left->parent = avln;
	}

	y->parent = avln->parent;

	if (avln->parent == NULL) {
		avlt->root = y;
	} else if (avln->value == avln->parent->left->value) {
		avln->parent->left = y;
	} else {
		avln->parent->right = y;
	}

	y->left = avln;
	avln->parent = y;

	adjust_height(avln);
	adjust_height(y);
}

// Performs a right rotation on the given node in the given tree
void rotate_right(AVLTree* avlt, AVLNode* avln) {
	AVLNode* x = avln->left;
	avln->left = x->right;

	if (x->right != NULL) {
		x->right->parent = avln;
	}

	x->parent = avln->parent;

	if (avln->parent == NULL) {
		avlt->root = x;
	} else if (avln->value == avln->parent->right->value) {
		avln->parent->right = x;
	} else {
		avln->parent->left = x;
	}

	x->right = avln;
	avln->parent = x;

	adjust_height(avln);
	adjust_height(x);
}

// Helper method to get the height of the given node
// Used to easily implement the required NULL-checks
int get_height(AVLNode* avln) {
	if (avln == NULL) {
		return 0;
	} else {
		return avln->height;
	}
}

// Helper method to get the height of the node to the left
// Used to easily implement the required NULL-checks
int get_left_height(AVLNode* avln) {
	if (avln == NULL || avln->left == NULL) {
		return 0;
	} else {
		return avln->left->height;
	}
}

// Helper method to get the height of the node to the right
// Used to easily implement the required NULL-checks
int get_right_height(AVLNode* avln) {
	if (avln == NULL || avln->right == NULL) {
		return 0;
	} else {
		return avln->right->height;
	}
}

// Balances the sub-tree below the given node
void balance(AVLTree* avlt, AVLNode* avln) {
	// check if the tree is left-heavy
	if (get_height(avln->left) > get_height(avln->right) + 1) {
		if (get_left_height(avln->left) < get_right_height(avln->left)) {
			rotate_left(avlt, avln->left);
		}
		rotate_right(avlt, avln);

	// check if the tree is right-heavy
	} else if (get_height(avln->right) > get_height(avln->left) + 1) {
		if (get_right_height(avln->right) < get_left_height(avln->right)) {
			rotate_right(avlt, avln->right);
		}
		rotate_left(avlt, avln);
	}
}

// Recursive method to insert a value into the tree, NOT to be used by the outside
void insert_value_recursive(AVLTree* avlt, AVLNode** avln, AVLNode* parent, int value) {
	// If the given node is NULL (i.e. the tree is empty or we have reached a leaf node)
	if (*avln == NULL) {
		*avln = malloc(sizeof(AVLNode));
		(*avln)->left = NULL;
		(*avln)->right = NULL;
		(*avln)->parent = parent;
		(*avln)->value = value;
		(*avln)->height = 1;

		avlt->size++;

	// If the the given value is lower than the currently viewed node, view the left child
	} else if (value < (*avln)->value) {
		insert_value_recursive(avlt, &(*avln)->left, *avln, value);

	// If the the given value is greater than the currently viewed node, view the right child
	} else if (value > (*avln)->value) {
		insert_value_recursive(avlt, &(*avln)->right, *avln, value);

	// If the the given value is equal to the currently viewed node, return because we don't allow double entries
	} else {
		return;
	}

	// Determine the height of the newly created node
	(*avln)->height = 1 + max(get_height((*avln)->left), get_height((*avln)->right));

	// Re-balance the node to check for possibly needed rotations
	balance(avlt, *avln);
}

// Inserts the given value into the tree
void insert_value(AVLTree* avlt, int value) {
	// Use the recursive function to allow the function to correctly position the newly created node
	insert_value_recursive(avlt, &avlt->root, NULL, value);
}

void traverse_in_order(AVLNode* avln) {
	if (avln != NULL) {
		// traverse the left sub-tree first
		traverse_in_order(avln->left);

		// print the current node's value
		printf("%d ", avln->value);

		// traverse the right sub-tree later
		traverse_in_order(avln->right);
	}
}

int main() {
	// create a new AVLTree
	AVLTree avlt;
	avlt.root = NULL;
	avlt.size = 0;

	insert_value(&avlt, 2);
	insert_value(&avlt, 5);
	insert_value(&avlt, 7);
	insert_value(&avlt, 3);
	insert_value(&avlt, 10);
	insert_value(&avlt, 1);
	insert_value(&avlt, 20);

	traverse_in_order((&avlt)->root);
	printf("\n");
}
